server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -527,17 +571,26 @@ public ClusterState execute(ClusterState currentState) {
             public void onFailure(String source, Exception e) {
                 logger.warn("failed to update snapshot state after node removal");
             }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                if (changed) {
```

Comment:
This is kind of lazy and obviously a better solution would be to simply track those snapshots that were completed in the update by adding them to a list or so to be processed here. Just didn't want to do that in this PR since it introduces quite a bit of complexity. But I figured just adding the `changed` flag here and in the other tasks was enough to illustrate the motivation for this change and is already a huge win in terms of not having to iterate all the shards on every CS application.